cmd/kubeadm/app/apis/kubeadm/types.go

```diff
@@ -24,68 +24,149 @@ import (
+	BootstrapToken BootstrapToken `json:"bootstrapToken"`
+	// Paths specifies locations to various important directories and files during initialization time
+	Paths MasterPaths `json:"paths"`
+
+	// KubernetesVersion controls which version to use for the control plane, the kube-proxy and which other features will be enabled
+	KubernetesVersion string `json:"kubernetesVersion"`
+	// NodeName specifies the name the Node API object will have in a later stage. Defaulting to os.Hostname(), but can be useful to set in a custom environment
+	NodeName string `json:"nodeName"`
+	// FeatureGates specifies which kubeadm-specific features to enable/disable on demand
+	FeatureGates map[string]bool `json:"featureGates"`
+	// ImageRepository specifies what container registry to pull control plane images from
+	ImageRepository string `json:"imageRepository"`
+}
 
-	Token    string
-	TokenTTL time.Duration
+// APIServer specifies how the api server component is set up and configured
+type APIServer struct {
+	// Image specifies which container image to use for running the api server. If empty, automatically populated by kubeadm using the ImageRepository and KubernetesVersion
+	Image string `json:"image"`
+	// ExtraArgs lets the user customize and/or override the arguments that are given to the api server
+	ExtraArgs map[string]string `json:"extraArgs"`
+	// ExtraHostPathVolumes lets the user add new hostPath volumes and volumeMounts as needed
+	ExtraHostPathVolumes []HostPathVolume `json:"extraHostPathVolumes"`
 
-	// SelfHosted enables an alpha deployment type where the apiserver, scheduler, and
-	// controller manager are managed by Kubernetes itself. This option is likely to
-	// become the default in the future.
-	SelfHosted bool
+	// AdvertiseAddress sets the address for the API server to advertise.
+	AdvertiseAddress string `json:"advertiseAddress"`
+	// BindPort sets the secure port for the API Server to bind to
+	BindPort int32 `json:"bindPort"`
+	// AuthorizationModes specifies which authorization modes to use for the api server
+	AuthorizationModes []string `json:"authorizationModes"`
+	// ServingCertExtraSANs sets extra Subject Alternative Names for the API Server signing cert
+	ServingCertExtraSANs []string `json:"servingCertExtraSANs"`
+}
 
-	APIServerExtraArgs         map[string]string
-	ControllerManagerExtraArgs map[string]string
-	SchedulerExtraArgs         map[string]string
+// ControllerManager specifies how the controller manager component is set up and configured
+type ControllerManager struct {
+	// Image specifies which container image to use for running the controller manager. If empty, automatically populated by kubeadm using the ImageRepository and KubernetesVersion
+	Image string `json:"image"`
+	// ExtraArgs lets the user customize and/or override the arguments that are given to the controller manager
+	ExtraArgs map[string]string `json:"extraArgs"`
```

Comment:
we avoid maps in the api as the order is undefined. Not sure whether this still causes trouble.